### PR TITLE
feat: Add CodeMirror editor to control panel

### PR DESCRIPTION
**Pull request by sum117**
# Code Mirror Editor for Control Panel
image

This PR adds CodeMirror to YAGPDB bot, by implementing the unused vendor package `CodeMirror`. There are style inconsistencies, but an admin told me it was alright, as users wouldn't care. I included a CSS file of a theme that I like, downloaded from their website directly. It should function as reference if someone ever wants to write the CSS to match the style of the highlight.js that is already included in this page [Permalink](https://github.com/botlabs-gg/yagpdb/blob/447a52732f5b901314397fe9079290e59d9f8d45/customcommands/assets/customcommands.html#L193).

## Why did I alter `spongebob.js`

I had to alter this file because I used a go template inside my code to access the keywords, and since a new textarea is created every time a button is clicked, I removed the jquery from `spongebob.js` and rewrote it in modern syntax, implementing the code mirror instance every click.

## About themes

I think it'll take a while for the team to decide which theme should be used, or maybe not, idk. In case you guys didn't like the theme I chose, just pick one from [here](https://codemirror.net/5/theme/) and i'll update the code.

My Discord is currently `shar_alter`.

### CODE REVIEW

File: customcommands/assets/customcommands-editcmd.html [hunk 1]
```diff
@@ -740,6 +740,146 @@ <h2 class="card-title">Custom Command Information</h2>
...
+<script src="/static/vendorr/codemirror/addon/mode/multiplex.js"></script>
+<script src="/static/vendorr/codemirror/addon/display/placeholder.js"></script>
+<style>
+    .CodeMirror {
+        width: 100% !important;
+        max-width: 95% !important;
+    }
+</style>
+<script type="text/javascript">
+    let keywords = ["for", "range", "if", "else", "template", "end", "true", "false", "nil", "gt", "lt", "len", "index"];
```
[the remark this answers is not in the document]
Comment (sum117) [reply customcommands/assets/customcommands-editcmd.html:756]: Hi.

I got those keywords from the HLJS implementation that already exists. You can check the contents in the permalink that I provided in the description of the PR. As for why they made the decision to not include the keywords you mentioned, I don't know.

[the remark this answers is not in the document]
Comment (galen8183) [reply customcommands/assets/customcommands-editcmd.html:756]: The existing implementation also includes all of [templates.StandardFuncMap](https://github.com/botlabs-gg/yagpdb/blob/447a52732f5b901314397fe9079290e59d9f8d45/common/templates/context.go#L31), see https://github.com/botlabs-gg/yagpdb/blob/447a52732f5b901314397fe9079290e59d9f8d45/customcommands/web.go#L253
The other missing block actions and functions from text/template were probably just an oversight and should be added here.

**Comment by Ranger-4297** [comment]
Currently loaded up this PR on my selfhost why implement a "syntax highlighting" that doesn't make a distinguishable barrier between functions and operators that are present in the code and random characters?
Obviously I assume this is one issue that will be fixed over time but I hope you at least noticed this when testing before submitting the PR
image

**Comment by sum117** [comment]
Yes, @Ranger-4297 .

That distinction can be added easily, and I did test it.

Judging by the image I included and comparing it to what existed before, I'd say it's a lot more helpful than just using "mode": "go", for example (it's what you'd get for using ```go <code>``` in discord, for reference). 

I'm aware improvements can be made, but this isn't broken.

**Comment by Ranger-4297** [comment]
Ok, cool. Just glad we're on the same page
I think it's just as important as having this highlighting to be able to distinguish between what is what
I don't mind helping out provided you can give me a pointer on what, either

File: customcommands/assets/customcommands-editcmd.html [hunk 2]
```diff
@@ -740,6 +740,146 @@ <h2 class="card-title">Custom Command Information</h2>
...
+<link rel="stylesheet" href="/static/vendorr/codemirror/lib/codemirror.css">
+<link rel="stylesheet" href="/static/vendorr/codemirror/theme/ayu-mirage.css">
+<script src="/static/vendorr/codemirror/lib/codemirror.js"></script>
+<script src="/static/vendorr/codemirror/mode/markdown/markdown.js"></script>
+<script src="/static/vendorr/codemirror/addon/mode/multiplex.js"></script>
+<script src="/static/vendorr/codemirror/addon/display/placeholder.js"></script>
+<style>
+    .CodeMirror {
+        width: 100% !important;
+        max-width: 95% !important;
```
Comment (galen8183) [inline customcommands/assets/customcommands-editcmd.html:752]:
```diff
+       resize: vertical;
```
This allows resizing the editor like the standard textarea did previously.

Comment (sum117) [reply customcommands/assets/customcommands-editcmd.html:752]: Added, @galen8183

**Comment by sum117** [comment]
Please, @Ranger-4297 , try it now:

image

File: frontend/static/js/joelFuncs.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
+// https://github.com/jo3-l/yagfuncdata
+var joelFuncs = {
```
Comment (galen8183) [inline frontend/static/js/joelFuncs.js:2]:
```suggestion
var joelFuncs = {
var yagFuncs = {
```
This should be a more descriptive name IMO

**Pull request merged**